t/04-unit-handler-remap.t

```diff
@@ -0,0 +1,138 @@
+#open( STDOUT, '>>', $basedir . 't/tmp/misc.err' ) || die "Cannot open errlog [$!]";
+
+my $base_tester = Mail::Milter::Authentication::Tester::HandlerTester->new({
+    'prefix'   => $basedir . 't/config/handler/etc',
+    'zonedata' => '',
+    'handler_config' => {
+        'IPRev' => {},
+        'PTR' => {},
+        'LocalIP' => {},
+        'TrustedIP' => {},
+        'Auth' => {},
+    },
+});
+
+my $no_remap_tester = clone $base_tester;
+my $remap_tester = clone $base_tester;
+$remap_tester->{ 'authmilter' }->{ 'config' }->{ 'ip_map' } = {
+    '1.1.0.0/24' => {
+        'ip' => '1.2.0.1',
+        'helo' => 'ip_remapped.example.com',
+        'helo_map' => {
+            'matched_helo.example.com' => {
+                'ip' => '1.3.0.1',
+                'helo' => 'helo_remapped.example.com',
+            }
+        }
+    }
+};
+my $use_tester;
+
+$use_tester = clone $no_remap_tester;
+subtest no_remap_smtp => sub {
+    run_test_on( $use_tester, 'smtp', '1.4.0.1', 'sent_helo.example.com', '1.4.0.1', 'sent_helo.example.com' );
+};
+
+$use_tester = clone $no_remap_tester;
+subtest no_remap_milter => sub {
+    run_test_on( $use_tester, 'milter', '1.4.0.1', 'sent_helo.example.com', '1.4.0.1', 'sent_helo.example.com' );
+};
+
+$use_tester = clone $remap_tester;
+subtest ip_remap_no_hit_smtp => sub {
+    run_test_on( $use_tester, 'smtp', '1.4.0.1', 'sent_helo.example.com', '1.4.0.1', 'sent_helo.example.com' );
+};
+
+$use_tester = clone $remap_tester;
+subtest ip_remap_no_hit_milter => sub {
+    run_test_on( $use_tester, 'milter', '1.4.0.1', 'sent_helo.example.com', '1.4.0.1', 'sent_helo.example.com' );
+};
+
+$use_tester = clone $remap_tester;
+subtest ip_remap_ip_hit_smtp => sub {
+    run_test_on( $use_tester, 'smtp', '1.1.0.1', 'sent_helo.example.com', '1.2.0.1', 'ip_remapped.example.com' );
+};
+
+$use_tester = clone $remap_tester;
+subtest ip_remap_ip_hit_milter => sub {
+    run_test_on( $use_tester, 'milter', '1.1.0.1', 'sent_helo.example.com', '1.2.0.1', 'ip_remapped.example.com' );
+};
+
+$use_tester = clone $remap_tester;
+subtest ip_remap_helo_hit_smtp => sub {
+    run_test_on( $use_tester, 'smtp', '1.1.0.1', 'matched_helo.example.com', '1.3.0.1', 'helo_remapped.example.com' );
+};
+
+$use_tester = clone $remap_tester;
+subtest ip_remap_helo_hit_milter => sub {
+    # IP cannot remap in milter protocol, so will be the generic remap not the helo remap
+    run_test_on( $use_tester, 'milter', '1.1.0.1', 'matched_helo.example.com', '1.2.0.1', 'helo_remapped.example.com' );
+};
+
+done_testing();
+
+sub run_test_on {
+  my ( $tester, $order, $send_ip, $send_helo, $expect_ip, $expect_helo ) = @_;
+
+  my $host = 'nothing.example.com';
+  my $ip = Net::IP->new( $send_ip );
+  my $helo = $send_helo;
+  my $from = 'from_address.example.com';
+  my $to = 'to_address.example.com';
+
+  my $handler = $tester->{ 'authmilter' }->{ 'handler' }->{ '_Handler' };;
+
+  if ( $order eq 'smtp' ) {
```

Comment:
I am a bit confused by this, but it's probably because I don't understand the test framework all that well.

It seems potentially misleading to say "I know that in the SMTP world, we'll call callbacks in this order and in the Milter protocol we'll do them in this other order." It seems like these tests would be better if you said something like (this made-up code): 
```perl
my $handler = $tester->get_smtp_handler;  # or get_milter_handler
$handler->process_request({ $request });
check_output()
```

But maybe we can't do that because we need to check the internal state of the handler after remapping but before we do anything else. Just a thought!

Reply:
There is a run (off the top of my head) method for the tester, but that doesn't call the remapping methods, perhaps it should do that.
The ::Tester::HandlerTester mocks up enough of a stack to test handlers, but without needing a full Protocol object, and as this change is mainly in the Protocol level, and most handlers should not be aware of the remapping (other than it's result) it seemed easier to add explicit calls to the remapping methods in the order that the Protocol objects would call them than to put remapping support into the tester for this one edge case of testing.